### PR TITLE
imgpkg: init at 0.43.1

### PR DESCRIPTION
**Pull request by BenChand**
## Description of changes

Added a package.nix for imgpkg

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/im/imgpkg/package.nix
Comment (SigmaSquadron) [reply pkgs/by-name/im/imgpkg/package.nix:16]: Avoid `rec`!

Comment (BenChand) [reply pkgs/by-name/im/imgpkg/package.nix]: Ah good to know, thanks, will avoid in the future

**Comment by BenChand** [comment]
Downgraded to 0.42.1 as 0.43.0 uses `go >= 1.22.5` which `buildGoModule` isn't up to yet

File: pkgs/by-name/im/imgpkg/package.nix
Comment (SigmaSquadron) [inline pkgs/by-name/im/imgpkg/package.nix:25]:
```suggestion
  env.CGO_ENABLED = "0";
```

File: pkgs/by-name/im/imgpkg/package.nix
Comment (SigmaSquadron) [inline pkgs/by-name/im/imgpkg/package.nix:33]:
```suggestion
    mainProgram = "imgpkg";
    platforms <either [ "singleton-platform" ] or lib.platforms.listOfPlatforms>;
```

**Review by SigmaSquadron — COMMENTED** [review]
Please squash your commits based on the [Commit Guidelines](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions).

File: pkgs/by-name/im/imgpkg/package.nix
Comment (SigmaSquadron) [reply pkgs/by-name/im/imgpkg/package.nix]: @BenChand: whoops, that was my bad. I assumed it was an environment variable to be passed to the builder, but it looks like the Go framework handles it differently. Remove the `env.`.

Comment (BenChand) [reply pkgs/by-name/im/imgpkg/package.nix]: Ah no worries, will revert

File: pkgs/by-name/im/imgpkg/package.nix
Comment (BenChand) [reply pkgs/by-name/im/imgpkg/package.nix:33]: Added platforms for linux and darwin

File: pkgs/by-name/im/imgpkg/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/im/imgpkg/package.nix:19]:
```suggestion
buildGoModule rec {
  pname = "imgpkg";
  version = "0.42.1";

  src = fetchFromGitHub {
    owner = "carvel-dev";
    repo = "imgpkg";
    rev = "v${version}";
    hash = "sha256-n8v27FpFpx0PBu/cCHgq128+fAjSGTEdLCG/g7MHCVg";
  };
```

File: pkgs/by-name/im/imgpkg/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/im/imgpkg/package.nix:34]:
```suggestion
```
buildGoModule already defaults to this

File: pkgs/by-name/im/imgpkg/package.nix
Comment (BenChand) [reply pkgs/by-name/im/imgpkg/package.nix:19]: It was suggested here to avoid `rec` so now I'm unsure which one to go with 😄 

https://github.com/NixOS/nixpkgs/pull/336558#discussion_r1727435400

File: pkgs/by-name/im/imgpkg/package.nix
Comment (BenChand) [reply pkgs/by-name/im/imgpkg/package.nix:34]: oh nice

File: pkgs/by-name/im/imgpkg/package.nix
Comment (SigmaSquadron) [reply pkgs/by-name/im/imgpkg/package.nix:19]: `rec` is an antipattern, is it not?

Comment (BenChand) [reply pkgs/by-name/im/imgpkg/package.nix:19]: I can't see any info in the contribution guidelines for either avoiding or adhering to `rec` as a pattern. @SuperSandro2000 @SigmaSquadron can we agree on a style to approve? I'd quite like to get this merged so I can add some other missing packages

Comment (SigmaSquadron) [reply pkgs/by-name/im/imgpkg/package.nix:19]: In the interest of not nitpicking, I'd agree with either style. However, `rec` is [definitely on its way out](https://github.com/NixOS/nixpkgs/issues/315337), and shouldn't be recommended anymore. Ideally, one would use `finalAttrs` if the builder supports it, or `let...in` if not. As the official Nix documentation explicitly recommends against overusing `rec`, I believe that `let...in` or `finalAttrs` is the ideal solution here, but Sandro has far more review experience than me so I'll defer to his judgement.

Keep `rec`, for now. Once a greater consensus is reached, we can do a massive tree-wide to remove it.

**Review by SigmaSquadron — APPROVED** [review]
SGTM, but please squash the commits to follow the Committing Guidelines.

**Comment by BenChand** [comment]
Updated to imgpkg `0.43.1` as `buildGoModule` now supports that version of `go`

**Review by SigmaSquadron — APPROVED** [review]
Approval maintained.

**Comment by BenChand** [comment]
@NixOS/nixpkgs-merge-bot merge

**Review by Aleksanaa — APPROVED** [review]
Looks good overall, just one more thing. Please split into two commits:
1. maintainers: add benchand
2. imgpkg: init at 0.43.1

**Comment by BenChand** [comment]
Yeah that makes sense actually, will do

**Comment by BenChand** [comment]
https://github.com/NixOS/nixpkgs/pull/343979

**Pull request merged**
